Python/Product/PythonTools/PythonTools/Debugger/DebugAdapter/DebugAdapterAttachProcess.cs

```diff
@@ -0,0 +1,203 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Net;
+using System.Net.Sockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.PythonTools.Debugger.DebugAdapter;
+using Microsoft.PythonTools.Infrastructure;
+using Microsoft.PythonTools.Logging;
+using Microsoft.VisualStudio.Debugger.DebugAdapterHost.Interfaces;
+using Microsoft.VisualStudio.Debugger.Interop;
+
+namespace Microsoft.PythonTools.Debugger {
+    internal sealed class DebugAdapterAttachProcess : ITargetHostProcess, IDisposable {
+        private const int _debuggerConnectionTimeout = 60000; // 60 seconds. Can take a long time to connect
+        private DebugAdapterAttachProcessStream _stream;
+        private bool _debuggerConnected = false;
+        private Process _debugPyProcess = null;
+
+        private DebugAdapterAttachProcess(Process debugPyProcess = null) {
+            _debugPyProcess = debugPyProcess;
+            debugPyProcess.Exited += DebugPyProcess_Exited;
+            debugPyProcess.OutputDataReceived += DebugPyProcess_OutputDataReceived;
+            debugPyProcess.ErrorDataReceived += DebugPyProcess_OutputDataReceived;
+        }
+
+        private void DebugPyProcess_OutputDataReceived(object sender, DataReceivedEventArgs e) {
+            Debug.WriteLine($"DebugPy output: {e.Data}");
+        }
+
+        private void DebugPyProcess_Exited(object sender, EventArgs e) {
+            Debug.WriteLine($"DebugPy exited:");
+            Exited?.Invoke(this, EventArgs.Empty);
+        }
+
+        public static ITargetHostProcess LocalAttach(DebugAttachInfo debugAttachInfo, IAdapterLaunchInfo launchInfo, ITargetHostInterop targetInterop) {
+            // For the local case, we need to start the process ourselves
+            AD_PROCESS_ID adProcessId = new AD_PROCESS_ID();
+            adProcessId.ProcessIdType = (uint)enum_AD_PROCESS_ID.AD_PROCESS_ID_SYSTEM;
+            adProcessId.dwProcessId = (uint)launchInfo.AttachProcessId;
+            launchInfo.DebugPort.GetProcess(adProcessId, out var process);
```

Comment:
does GetProcess return something like success/fail or error?

Reply:
It returns an HR result. All of the debugger APIs are actually COM calls.

However if you have a port, you always have a process. It's exceptional (hence an exception) if there isn't one.

Reply:
If it's an HRESULT return from managed perspective, it should still check - I think there was a `RequireOk` helper method somewhere?